In [1]:
import numpy as np
import pandas as pd

## 1º Carregue em dataframes os conjuntos de dados referentes ao ano de 2018, 2019 e 2020. Corrija quaisquer erros que apareçam no carregamento dos arquivos. É possível acessar os arquivos compactados diretamente do site, sem precisar baixá-los?

In [2]:
arq2020 = pd.read_csv('https://arquivos.prf.gov.br/arquivos/index.php/s/jdDLrQIf33xXSCe/download', 
                      compression='zip', encoding = 'iso-8859-1', sep = ';')


arq2019 = pd.read_csv('https://arquivos.prf.gov.br/arquivos/index.php/s/kRBUylqz6DyQznN/download', 
                      compression='zip', encoding = 'iso-8859-1', sep = ';')


arq2018 = pd.read_csv('https://arquivos.prf.gov.br/arquivos/index.php/s/MaC6cieXSFACNWT/download', 
                      compression='zip', encoding = 'iso-8859-1', sep = ';')

#### Resposta: Sim, é possível. Ao utilizarmos a função "compression" podemos descompactar o arquivo ao informar o paramêtro, este sendo o tipo de compactação, que neste caso era "zip". 
##### Legenda:
        Acidentes em 2020 = arq2020
        Acidentes em 2019 = arq2019
        Acidentes em 2018 = arq2018

### OBS: Tratarei os nomes das colunas e após isso usarei "concat" para unir as três, já que todas possuem as mesmas variáveis.

In [3]:
arq2020r = arq2020.rename(columns = {'id' : 'ID', 'data_inversa' : 'DATA INVERSA', 'dia_semana' : 'DIA DA SEMANA', 'horario' : 'HORÁRIO',
                                    'uf' : "ESTADO", 'br' : 'ESTRADA', 'km' : 'KM', 'municipio' :'MUNICIPÍO', 'causa_acidente' : 'CAUSA DO ACIDENTE', 
                                    'tipo_acidente' : 'TIPO DO ACIDENTE', 'classificacao_acidente' : 'CLASSIFICAÇÃO DO ACIDENTE', 'fase_dia' : 'FASE DO DIA', 
                                     'sentido_via' : 'SENTIDO DA VIA', 'condicao_metereologica' : 'CONDIÇÃO', 'tipo_pista' : 'TIPO DE PISTA', 'tracado_via' : 'TRAÇADO DA VIA', 
                                     'uso_solo' : 'USO DO SOLO', 'pessoas' : 'PESSOAS ENVOLVIDAS', 'mortos' : 'MORTOS', 'feridos_leves' : 'FERIDOS LEVES', 
                                     'feridos_graves' : 'FERIDOS GRAVES', 'ilesos' : 'ILESOS', 'ignorados' : 'IGNORADOS', 'feridos' : 'FERIDOS', 
                                     'veiculos' : 'VEÍCULOS', 'latitude' : 'LATITUDE', 'longitude' : 'LONGITUDE', 'regional' : 'REGIONAL', 
                                     'delegacia' : 'DELEGACIA', 'uop' : 'UOP'})

arq2019r = arq2019.rename(columns = {'id' : 'ID', 'data_inversa' : 'DATA INVERSA', 'dia_semana' : 'DIA DA SEMANA', 'horario' : 'HORÁRIO',
                                    'uf' : "ESTADO", 'br' : 'ESTRADA', 'km' : 'KM', 'municipio' :'MUNICIPÍO', 'causa_acidente' : 'CAUSA DO ACIDENTE', 
                                    'tipo_acidente' : 'TIPO DO ACIDENTE', 'classificacao_acidente' : 'CLASSIFICAÇÃO DO ACIDENTE', 'fase_dia' : 'FASE DO DIA', 
                                     'sentido_via' : 'SENTIDO DA VIA', 'condicao_metereologica' : 'CONDIÇÃO', 'tipo_pista' : 'TIPO DE PISTA', 'tracado_via' : 'TRAÇADO DA VIA', 
                                     'uso_solo' : 'USO DO SOLO', 'pessoas' : 'PESSOAS ENVOLVIDAS', 'mortos' : 'MORTOS', 'feridos_leves' : 'FERIDOS LEVES', 
                                     'feridos_graves' : 'FERIDOS GRAVES', 'ilesos' : 'ILESOS', 'ignorados' : 'IGNORADOS', 'feridos' : 'FERIDOS', 
                                     'veiculos' : 'VEÍCULOS', 'latitude' : 'LATITUDE', 'longitude' : 'LONGITUDE', 'regional' : 'REGIONAL', 
                                     'delegacia' : 'DELEGACIA', 'uop' : 'UOP'})

arq2018r = arq2018.rename(columns = {'id' : 'ID', 'data_inversa' : 'DATA INVERSA', 'dia_semana' : 'DIA DA SEMANA', 'horario' : 'HORÁRIO',
                                    'uf' : "ESTADO", 'br' : 'ESTRADA', 'km' : 'KM', 'municipio' :'MUNICIPÍO', 'causa_acidente' : 'CAUSA DO ACIDENTE', 
                                    'tipo_acidente' : 'TIPO DO ACIDENTE', 'classificacao_acidente' : 'CLASSIFICAÇÃO DO ACIDENTE', 'fase_dia' : 'FASE DO DIA', 
                                     'sentido_via' : 'SENTIDO DA VIA', 'condicao_metereologica' : 'CONDIÇÃO', 'tipo_pista' : 'TIPO DE PISTA', 'tracado_via' : 'TRAÇADO DA VIA', 
                                     'uso_solo' : 'USO DO SOLO', 'pessoas' : 'PESSOAS ENVOLVIDAS', 'mortos' : 'MORTOS', 'feridos_leves' : 'FERIDOS LEVES', 
                                     'feridos_graves' : 'FERIDOS GRAVES', 'ilesos' : 'ILESOS', 'ignorados' : 'IGNORADOS', 'feridos' : 'FERIDOS', 
                                     'veiculos' : 'VEÍCULOS', 'latitude' : 'LATITUDE', 'longitude' : 'LONGITUDE', 'regional' : 'REGIONAL', 
                                     'delegacia' : 'DELEGACIA', 'uop' : 'UOP'})

arqtotal = pd.concat([arq2018r, arq2019r, arq2020r])

In [4]:
arqtotal

ID DATA INVERSA  DIA DA SEMANA   HORÁRIO ESTADO  ESTRADA     KM  \
0       99973.0   2018-01-01  segunda-feira  00:20:00     RJ    116.0  303,5   
1       99976.0   2018-01-01  segunda-feira  00:40:00     SC    282.0    0,4   
2       99977.0   2018-01-01  segunda-feira  00:30:00     RJ    493.0      1   
3       99981.0   2018-01-01  segunda-feira  01:15:00     RS    386.0    134   
4       99982.0   2018-01-01  segunda-feira  00:20:00     RS    293.0  151,7   
...         ...          ...            ...       ...    ...      ...    ...   
63543  364615.0   2020-03-20    sexta-feira  09:00:00     RN    304.0  171,8   
63544  364810.0   2020-09-12         sábado  23:50:00     BA    116.0   11,5   
63545  364817.0   2020-10-30    sexta-feira  10:30:00     BA    116.0   14,3   
63546  364835.0   2020-09-20        domingo  23:00:00     BA    116.0      8   
63547  364903.0   2020-10-31         sábado  02:10:00     BA    116.0      3   

                  MUNICIPÍO                   CAUSA DO ACIDENTE  \
0                   RESENDE                   Condutor Dormindo   
1             FLORIANOPOLIS  Não guardar distância de segurança   
2                  ITABORAI              Ultrapassagem Indevida   
3                   SARANDI                  Ingestão de Álcool   
4                  CANDIOTA         Falta de Atenção à Condução   
...                     ...                                 ...   
63543               ANGICOS         Falta de Atenção à Condução   
63544  VITORIA DA CONQUISTA                  Ingestão de Álcool   
63545  VITORIA DA CONQUISTA         Falta de Atenção à Condução   
63546  VITORIA DA CONQUISTA             Velocidade Incompatível   
63547  VITORIA DA CONQUISTA                    Animais na Pista   

                   TIPO DO ACIDENTE  ... FERIDOS GRAVES ILESOS IGNORADOS  \
0         Saída de leito carroçável  ...              4      1         0   
1                  Colisão traseira  ...              2      1         0   
2                   Colisão frontal  ...              0      3         1   
3               Colisão transversal  ...              0      2         0   
4         Saída de leito carroçável  ...              1      0         0   
...                             ...  ...            ...    ...       ...   
63543              Colisão traseira  ...              1      2         0   
63544               Colisão lateral  ...              0      2         0   
63545               Colisão lateral  ...              0      2         0   
63546  Queda de ocupante de veículo  ...              0      0         0   
63547       Atropelamento de Animal  ...              1      0         0   

      FERIDOS VEÍCULOS      LATITUDE     LONGITUDE  REGIONAL  DELEGACIA  \
0           4        1     -22,46937     -44,44705     SR-RJ     DEL5/7   
1           2        2  -27,59971687  -48,57565694     SR-SC     DEL8/1   
2           1        3    -22,763901    -42,927532     SR-RJ     DEL5/4   
3           0        2    -27,953636    -52,916374     SR-RS    DEL9/14   
4           1        1    -31,395214  -53,78391174     SR-RS    DEL9/11   
...       ...      ...           ...           ...       ...        ...   
63543       2        4   -5,68711053    -36,419458     SR-RN      UOP03   
63544       1        2  -14,87308214  -40,87141514     SR-BA      UOP01   
63545       2        2  -14,89597653  -40,86098671     SR-BA      UOP01   
63546       1        1  -14,84732305  -40,88634968     SR-BA      UOP01   
63547       1        1    -14,824848    -40,861311     SR-BA      UOP01   

            UOP  
0      UOP03/RJ  
1      UOP01/SC  
2      UOP02/RJ  
3      UOP01/RS  
4      UOP03/RS  
...         ...  
63543     DEL01  
63544     DEL08  
63545     DEL08  
63546     DEL08  
63547     DEL08  

[200289 rows x 30 columns]

## 2º Explore os conjuntos de dados, respondendo as questões:

### a) Quantos acidentes ocorreram em cada ano?

In [5]:
arq2018r['ID'].count()
arq2019r['ID'].count()
arq2020r['ID'].count()
soma = arq2020r['ID'].count()+arq2019r['ID'].count()+arq2018r['ID'].count()
dados = { 'TOTAL' : [arq2018r['ID'].count(),arq2019r['ID'].count(),arq2020r['ID'].count(), soma]}
indice = ['ACIDENTES REGISTRADOS EM 2018', 'ACIDENTES REGISTRADOS EM 2019', 'ACIDENTES REGISTRADOS EM 2020', 'SOMA']
ocorrencias = pd.DataFrame(dados, indice)
ocorrencias

TOTAL
ACIDENTES REGISTRADOS EM 2018   69295
ACIDENTES REGISTRADOS EM 2019   67446
ACIDENTES REGISTRADOS EM 2020   63548
SOMA                           200289

#### COMENTÁRIOS: O críterio para contar a quantidade de acidentes em cada ano foi a coluna "ID", já que é a primeira informação registrada pelo sistema quando um acidente ocorre. Por consequência é o número de linhas do nosso DataFrame.

 ### b) Quantas variáveis cada conjunto de dados registra?

In [6]:
dados = {'Nº VARIÁVEIS' : [len(arq2018r.columns),len(arq2019r.columns),len(arq2020r.columns)]}
indice = ['2018', '2019', '2020']
variaveis = pd.DataFrame(dados, indice)
variaveis

Nº VARIÁVEIS
2018            30
2019            30
2020            30

#### COMENTÁRIOS: As variáveis de um DataFrame nada mais são que suas colunas, sendo assim, usando o nome de um data frame, arq2018r por exemplo, com a função ".columns", obteremos as próprias colunas deste dataframe, logo, suas variáveis. Desta forma, podemos usar a função "len" para contar o numero de colunas.

### C) Quais as cinco cidades brasileiras onde mais ocorreram acidentes em rodovias federais?


In [7]:
municipios2020 = arq2020r['MUNICIPÍO'].value_counts().to_frame().head()
municipios2019 = arq2019r['MUNICIPÍO'].value_counts().to_frame().head()
municipios2018 = arq2018r['MUNICIPÍO'].value_counts().to_frame().head()
municipiostotal = arqtotal['MUNICIPÍO'].value_counts().to_frame().head()
municipiosgeral = pd.concat([municipios2020, municipios2019, municipios2018, municipiostotal], 
                            keys = ['2020', '2019', '2018', 'GERAL'])
municipiosgeral

MUNICIPÍO
2020  BRASILIA        1040
      CURITIBA         950
      GUARULHOS        639
      SAO JOSE         629
      PALHOCA          593
2019  CURITIBA        1094
      BRASILIA        1090
      SAO JOSE         819
      GUARULHOS        717
      SERRA            633
2018  CURITIBA        1019
      SAO JOSE         874
      BRASILIA         867
      GUARULHOS        755
      PALHOCA          688
GERAL CURITIBA        3063
      BRASILIA        2997
      SAO JOSE        2322
      GUARULHOS       2111
      PALHOCA         1887

#### COMENTÁRIOS: Considerando os 3 anos, as cidades com o maior número de acidentes foram: Curitiba, Brasília, São José, Guarulhos e Palhoça. Para contar esses dados eu utilizei a função "value_counts()" que nada mais faz do que contar quantas vezes valores se repetiram. Como nós queremos especificamente saber quais foram as cidades, eu filtrei apenas a coluna dos municípios, além disso usei "head()", já que nós queremos apenas as cinco cidades com maiores ocorrências. Após isso usei "concat" para agrupar os resultados em um único dataframe, e usei a função "Keys" para separar os dados por ano.

### D) Quantos acidentes com feridos graves aconteceram na Paraíba em 2019?

In [8]:
acidentesPB = arq2019r[arq2019r['ESTADO'] == 'PB']
acidentesgravesPB = acidentesPB[(acidentesPB['FERIDOS GRAVES'] >= 1)]
dados = {'TOTAL' : acidentesgravesPB['ID'].count()}
acidentespb = pd.DataFrame(dados, ['ACIDENTES COM FERIDOS GRAVES PB'])
acidentespb

TOTAL
ACIDENTES COM FERIDOS GRAVES PB    414

#### COMENTÁRIOS: Para alcançar este resultado, primeiro eu criei uma variável que carrega a parte do dataframe em que a coluna dos estados só contém resultados "PB" (Paraíba), após isso usei um argumento condicional para filtrar apenas os valores que nós buscamos, que é o de acidentes com pelo menos uma pessoa com ferimentos graves, após isso gerei um data frame com essa informação.

## 3º Para cada indicação abaixo, construa um novo dataframe, salvando-o com a terminação .csv.

### A) Ranking de acidentes por estado para todos os anos (um só arquivo combinando a informação dos três anos);


In [9]:
a = arq2018r['ESTADO'].value_counts().to_frame().reset_index().rename(columns = {'index': 'ESTADO' ,'ESTADO': 'Nº OCORRÊNCIAS'})
b = arq2019r['ESTADO'].value_counts().to_frame().reset_index().rename(columns = {'index': 'ESTADO' ,'ESTADO': 'Nº OCORRÊNCIAS'})
c = arq2020r['ESTADO'].value_counts().to_frame().reset_index().rename(columns = {'index': 'ESTADO' ,'ESTADO': 'Nº OCORRÊNCIAS'})
d = arqtotal['ESTADO'].value_counts().to_frame().reset_index().rename(columns = {'index': 'ESTADO' ,'ESTADO': 'Nº OCORRÊNCIAS'})
rankinggeral = pd.concat([a, b, c, d], keys = ['2018', '2019', '2020', 'GERAL'],axis = 1)
print('RANKING GERAL DE OCORRÊNCIAS EM TODOS OS ANOS')
rankinggeral

RANKING GERAL DE OCORRÊNCIAS EM TODOS OS ANOS


2018                  2019                  2020                 GERAL  \
   ESTADO Nº OCORRÊNCIAS ESTADO Nº OCORRÊNCIAS ESTADO Nº OCORRÊNCIAS ESTADO   
0      MG           9066     MG           8720     MG           8374     MG   
1      SC           8497     SC           8428     SC           7220     SC   
2      PR           7952     PR           7708     PR           7190     PR   
3      RJ           4575     RJ           4606     RJ           4236     RJ   
4      SP           4516     RS           4595     RS           4176     RS   
5      RS           4445     SP           4377     SP           4043     SP   
6      BA           3555     BA           3471     BA           3456     BA   
7      GO           3464     GO           3354     GO           3228     GO   
8      PE           2715     ES           2842     PE           2560     ES   
9      ES           2642     PE           2736     ES           2534     PE   
10     MT           2631     MT           2137     MT           2195     MT   
11     CE           1697     CE           1694     CE           1571     CE   
12     RO           1629     PB           1564     MS           1540     RO   
13     PB           1588     RO           1533     RO           1486     PB   
14     MS           1530     RN           1437     PB           1401     MS   
15     RN           1416     MS           1435     RN           1337     RN   
16     PI           1371     PI           1346     PI           1203     PI   
17     MA           1272     DF           1090     MA           1154     MA   
18     PA           1260     MA           1075     DF           1040     PA   
19     DF            867     PA            817     PA            963     DF   
20     AL            691     AL            629     AL            628     AL   
21     TO            589     SE            612     TO            595     SE   
22     SE            559     TO            520     SE            575     TO   
23     RR            236     AC            254     AC            300     AC   
24     AC            235     RR            237     RR            250     RR   
25     AP            173     AP            128     AP            155     AP   
26     AM            124     AM            101     AM            138     AM   

                   
   Nº OCORRÊNCIAS  
0           26160  
1           24145  
2           22850  
3           13417  
4           13216  
5           12936  
6           10482  
7           10046  
8            8018  
9            8011  
10           6963  
11           4962  
12           4648  
13           4553  
14           4505  
15           4190  
16           3920  
17           3501  
18           3040  
19           2997  
20           1948  
21           1746  
22           1704  
23            789  
24            723  
25            456  
26            363

In [10]:
rankinggeral.to_csv('rankinggeral.csv', index = True)
#Para executar use:
# pd.read_csv('rankinggeral.csv')

#### COMENTÁRIOS: Para fazer esse ranking eu usei a função,  "value_counts()", utilizada anteriormente (2º letra C). Após isso concatenei os dataframes horizontalmente, usei concat, pois por ser um ranking era importante que os nomes dos estados se repetissem, já que eles mudam de posição ao longo dos anos e nos resultados gerais. Para converter em "CSV" usei a função "to_csv" e mantive o índice.

## B)  Acidentes por dia da semana para todos os anos (um só arquivo combinando a informação dos três anos).

In [11]:
a = arq2018r['DIA DA SEMANA'].value_counts().to_frame()
b = arq2019r['DIA DA SEMANA'].value_counts().to_frame()
c = arq2020r['DIA DA SEMANA'].value_counts().to_frame()
d = arqtotal['DIA DA SEMANA'].value_counts().to_frame()
acidentessemana = pd.concat([a,b,c,d],axis=1)
acidentessemana.columns = ['2018', '2019', '2020', 'TOTAL']
acidentessemana

2018   2019   2020  TOTAL
domingo        11283  11298  10420  33001
sábado         11239  11099  10667  33005
sexta-feira    10632  10381   9764  30777
segunda-feira   9422   9134   8519  27075
quinta-feira    9230   8755   8244  26229
quarta-feira    8824   8374   8155  25353
terça-feira     8665   8405   7779  24849

In [12]:
acidentessemana.to_csv('acidentessemana.csv', index = True)
#Para executar use:
# pd.read_csv('acidentessemana.csv')

#### COMENTÁRIOS:  Nesta situação não era necessário fazer um ranking e sim a quantidade de acidentes nos dias da semana, eu apenas fiz concat sem a necessidade de repetir os dias da semana. Após isso usei o mesmo método da questão anterior para criar um arquivo "CSV".

## 4º Com relação a ocorrência dos acidentes, responda:

### A) Qual a causa mais frequente e a mais rara de acidentes registrados nos três anos?

In [13]:
comum = arqtotal['CAUSA DO ACIDENTE'].value_counts().to_frame().head(1)
rara = arqtotal['CAUSA DO ACIDENTE'].value_counts().to_frame().tail(12)
combinacao = pd.concat([comum,rara])
combinacao

CAUSA DO ACIDENTE
Falta de Atenção à Condução                                     73231
Pedestre cruzava a pista fora da faixa                              1
Acumulo de areia ou detritos sobre o pavimento                      1
Acumulo de óleo sobre o pavimento                                   1
Estacionar ou parar em local proibido                               1
Ausência de sinalização                                             1
Acostamento em desnível                                             1
Condutor usando celular                                             1
Fumaça                                                              1
Curva acentuada                                                     1
Frear bruscamente                                                   1
Ingestão de álcool ou de substâncias psicoativa...                  1
Pista esburacada                                                    1

#### COMENTÁRIOS: Mais uma vez usando "value_counts()" eu peguei a ocorrência mais comum que foi: "Falta de Atenção à Condução", da mesma forma fiz para a ocorrência mais rara, porém encontrei 12 valores igualmente raros nesses 3 anos, por isso usei a função "tail" que coleta os valores do final do data frame, usando o argumento 12, para exibir todas as mais raras. Por fim, concatenei os valores usando "concat".

## B) Qual a proporção de pessoas ilesas e de feridos graves por mês em cada um dos anos considerados? (proporção = numero de ilesos ou feridos graves/ número total de pessoas envolvidas no acidente);

In [14]:
#2018
#JANEIRO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-01')]
pgraves2018_01 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_01 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#FEVEREIRO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-02')]
pgraves2018_02 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_02 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#MARÇO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-03')]
pgraves2018_03 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_03 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#ABRIL
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-04')]
pgraves2018_04 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_04 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#MAIO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-05')]
pgraves2018_05 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_05 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#JUNHO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-06')]
pgraves2018_06 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_06 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#JULHO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-07')]
pgraves2018_07 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_07 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#AGOSTO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-08')]
pgraves2018_08 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_08 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#SETEMBRO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-09')]
pgraves2018_09 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_09 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#OUTUBRO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-10')]
pgraves2018_10 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_10 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#NOVEMBRO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-11')]
pgraves2018_11 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_11 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#DEZEMBRO
acdmes2018 = arq2018r[arq2018r['DATA INVERSA'].str.contains('2018-12')]
pgraves2018_12 = acdmes2018['FERIDOS GRAVES'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()
pilesos2018_12 = acdmes2018['ILESOS'].sum()/acdmes2018['PESSOAS ENVOLVIDAS'].sum()


#-----------------------------------------------------------------------------------

#2019
#JANEIRO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-01')]
pgraves2019_01 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_01 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#FEVEREIRO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-02')]
pgraves2019_02 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_02 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#MARÇO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-03')]
pgraves2019_03 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_03 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#ABRIL
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-04')]
pgraves2019_04 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_04 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#MAIO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-05')]
pgraves2019_05 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_05 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#JUNHO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-06')]
pgraves2019_06 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_06 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#JULHO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-07')]
pgraves2019_07 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_07 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#AGOSTO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-08')]
pgraves2019_08 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_08 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#SETEMBRO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-09')]
pgraves2019_09 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_09 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#OUTUBRO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-10')]
pgraves2019_10 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_10 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#NOVEMBRO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-11')]
pgraves2019_11 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_11 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#DEZEMBRO
acdmes2019 = arq2019r[arq2019r['DATA INVERSA'].str.contains('2019-12')]
pgraves2019_12 = acdmes2019['FERIDOS GRAVES'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()
pilesos2019_12 = acdmes2019['ILESOS'].sum()/acdmes2019['PESSOAS ENVOLVIDAS'].sum()


#-----------------------------------------------------------------------------------

#2020
#JANEIRO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-01')]
pgraves2020_01 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_01 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#FEVEREIRO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-02')]
pgraves2020_02 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_02 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#MARÇO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-03')]
pgraves2020_03 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_03 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#ABRIL
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-04')]
pgraves2020_04 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_04 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#MAIO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-05')]
pgraves2020_05 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_05 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#JUNHO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-06')]
pgraves2020_06 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_06 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#JULHO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-07')]
pgraves2020_07 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_07 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#AGOSTO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-08')]
pgraves2020_08 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_08 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#SETEMBRO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-09')]
pgraves2020_09 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_09 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#OUTUBRO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-10')]
pgraves2020_10 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_10 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#NOVEMBRO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-11')]
pgraves2020_11 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_11 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()


#DEZEMBRO
acdmes2020 = arq2020r[arq2020r['DATA INVERSA'].str.contains('2020-12')]
pgraves2020_12 = acdmes2020['FERIDOS GRAVES'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()
pilesos2020_12 = acdmes2020['ILESOS'].sum()/acdmes2020['PESSOAS ENVOLVIDAS'].sum()

In [15]:
a = {'ILESOS':[pilesos2020_01, pilesos2020_02, pilesos2020_03, pilesos2020_04, pilesos2020_05, 
               pilesos2020_06, pilesos2020_07, pilesos2020_08, pilesos2020_09, pilesos2020_10, 
               pilesos2020_11, pilesos2020_12],
    'GRAVES': [pgraves2020_01, pgraves2020_02, pgraves2020_03, pgraves2020_04, pgraves2020_05, 
               pgraves2020_06, pgraves2020_07, pgraves2020_08, pgraves2020_09, pgraves2020_10, 
               pgraves2020_11, pgraves2020_12]}


b = {'ILESOS':[pilesos2019_01, pilesos2019_02, pilesos2019_03, pilesos2019_04, pilesos2019_05, 
               pilesos2019_06, pilesos2019_07, pilesos2019_08, pilesos2019_09, pilesos2019_10, 
               pilesos2019_11, pilesos2019_12],
    'GRAVES': [pgraves2019_01, pgraves2019_02, pgraves2019_03, pgraves2019_04, pgraves2019_05, 
               pgraves2019_06, pgraves2019_07, pgraves2019_08, pgraves2019_09, pgraves2019_10, 
               pgraves2019_11, pgraves2019_12]}


c = {'ILESOS':[pilesos2018_01, pilesos2018_02, pilesos2018_03, pilesos2018_04, pilesos2018_05, 
               pilesos2018_06, pilesos2018_07, pilesos2018_08, pilesos2018_09, pilesos2018_10, 
               pilesos2018_11, pilesos2018_12],
    'GRAVES': [pgraves2018_01, pgraves2018_02, pgraves2018_03, pgraves2018_04, pgraves2018_05, 
               pgraves2018_06, pgraves2018_07, pgraves2018_08, pgraves2018_09, pgraves2018_10, 
               pgraves2018_11, pgraves2018_12]}


AcidentespMes2020 = pd.DataFrame(a)
AcidentespMes2019 = pd.DataFrame(b)
AcidentespMes2018 = pd.DataFrame(c)
AcidentesPorAnoeMes = pd.concat([AcidentespMes2018, AcidentespMes2019, AcidentespMes2020], keys = ['2018','2019','2020'], axis = 1)
AcidentesPorAnoeMes.index = ['JANEIRO', 'FEVEREIRO', 'MARÇO', 'ABRIL', 'MAIO', 'JUNHO', 
                             'JULHO', 'AGOSTO', 'SETEMBRO', 'OUTUBRO', 'NOVEMBRO', 'DEZEMBRO']
AcidentesPorAnoeMes.index.name = 'MÊS'

In [16]:
AcidentesPorAnoeMes

2018                2019                2020          
             ILESOS    GRAVES    ILESOS    GRAVES    ILESOS    GRAVES
MÊS                                                                  
JANEIRO    0.513291  0.090609  0.424224  0.113264  0.415565  0.107463
FEVEREIRO  0.498256  0.092191  0.424104  0.106789  0.420011  0.111356
MARÇO      0.481535  0.101853  0.433193  0.114599  0.401021  0.112575
ABRIL      0.430159  0.112774  0.415685  0.105970  0.376267  0.117949
MAIO       0.422230  0.114457  0.415162  0.120750  0.401330  0.115942
JUNHO      0.425097  0.116395  0.423757  0.121617  0.423283  0.115936
JULHO      0.437509  0.110295  0.436815  0.116670  0.424321  0.108617
AGOSTO     0.427051  0.106371  0.413084  0.116683  0.423414  0.119168
SETEMBRO   0.420665  0.115453  0.421064  0.110656  0.416016  0.117388
OUTUBRO    0.430178  0.112645  0.418922  0.119805  0.411006  0.119716
NOVEMBRO   0.433356  0.115405  0.425812  0.111047  0.407042  0.120385
DEZEMBRO   0.432916  0.108103  0.422265  0.114405  0.412292  0.121993

#### COMENTÁRIOS: Primeiramente, como não temos uma coluna "mês" e a única referência que temos ao mês é a coluna "DATA INVERTIDA" eu usei a função "contains" para localizar o mês a partir de um fragmento da string das linhas do data frame, após isso criei uma variável que recebeu essa filtragem, a partir disso, filtrei os feridos graves e os ilesos dos acidentes e dividi pela coluna "PESSOAS ENVOLVIDAS", adquirindo assim as suas proporções, com isso criei 3 data frames que receberam esses dados mês a mês e os concatenei os separando com "keys" por ano. Depois de montado, mudei os índices pra que fossem os meses e mudei o nome do índice para ser: "MÊS".
#### Analisando os resultados: Ao analisar os dados chegamos a conclusão de que em média 10% dos envolvidos em acidentes se ferem gravemente, por outro lado por volta de 40% dos envolvidos saem ilesos. 

## C) Mostre, reorganizando o conjunto de dados, se a pandemia conseguiu diminuir ou não a incidência de acidentes nas rodovias federais.


In [17]:
media = (arq2018['id'].count()+arq2019['id'].count())/2
indice = ['2018', '2019', '2020', 'MÉDIA PRE-PANDEMIA']
dados = {'NÚMERO DE ACIDENTES' : [arq2018['id'].count(), arq2019['id'].count(), arq2020['id'].count(), media]}
nacidentes = pd.DataFrame(dados, index = indice)
nacidentes.index.name = 'ANOS'
nacidentes

NÚMERO DE ACIDENTES
ANOS                                   
2018                            69295.0
2019                            67446.0
2020                            63548.0
MÉDIA PRE-PANDEMIA              68370.5

##### Redução de 4822 acidentes. -7% Em Relação a Média de Anos Anteriores.
#### COMENTÁRIOS: Para realizar este cálculo eu fiz a média dos anos pré-pandemia do nosso conjunto de dados, que foram os anos 2019 e 2018, o total foi de: 68370.5 ocorrências, ao comparar esse valor com as ocorrências de 2020, temos uma diminuição de 4822 acidentes, uma redução de 7%, ao olhar em um escopo maior não é uma redução enorme, porém se tratando de vidas já é um resultado muito relevante.